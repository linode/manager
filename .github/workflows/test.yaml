# This workflow will do a clean install of node dependencies, build the source code and run tests across different versions of node
# For more information see: https://help.github.com/actions/language-and-framework-guides/using-nodejs-with-github-actions

name: SDK Tests

on:
  push:
    branches: [ master, M3-github-actions ]

jobs:
  build:

    runs-on: ubuntu-latest

    strategy:
      matrix:
        node-version: [10.x]

    steps:
    - uses: actions/checkout@v2
    - name: Use Node.js ${{ matrix.node-version }}
      uses: actions/setup-node@v1
      with:
        node-version: ${{ matrix.node-version }}
    - run: npm install -g yarn
    - run: yarn install:all
    - name: Unit tests
      run: yarn workspace linode-js-sdk run test