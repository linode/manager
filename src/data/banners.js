export const testApiBanner = {
  entity: {
    id: 1234,
    type: 'ticket',
    label: 'testing ticket',
  },
  type: 'important_ticket',
  when: null,
};

export const testBillingBanner = {
  ...testApiBanner,
  entity: {
    id: 1235,
    label: 'Please update your billing information.',
  },
};

export const testAbuseBanner = {
  ...testApiBanner,
  entity: {
    id: 1235,
    label: 'ToS Violation',
  },
  type: 'abuse_ticket',
};

export const testAbuseBanner2 = {
  ...testAbuseBanner,
  entity: {
    id: 1236,
  },
};

export const importantBanner = [
  testApiBanner,
];

export const importantBanners = [
  testApiBanner,
  testBillingBanner,
];

export const abuseBanner = [
  testAbuseBanner,
];

export const abuseBanners = [
  testAbuseBanner,
  testAbuseBanner2,
];

export const banners = [
  testApiBanner,
  testBillingBanner,
  testAbuseBanner,
];
