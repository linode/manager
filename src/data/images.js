export const apiTestImage = {
  id: 'linode/arch2016.05',
  recommended: true,
  vendor: 'Arch',
  label: 'Arch Linux 2016.05',
  size: 550,
  created: '2009-01-01T00:00:00',
  description: 'Test Image',
  type: 'manual',
  created_by: 'linode',
  is_public: true,
  deprecated: false,
};

export const testLinodeImage = {
  ...apiTestImage,
  _polling: false,
};

export const publicImages = {
  'linode/arch2016.05': testLinodeImage,
  'linode/debian7': {
    ...testLinodeImage,
    id: 'linode/debian7',
    vendor: 'Debian',
    label: 'Debian 7',
  },
  'linode/debian8.1': {
    ...testLinodeImage,
    id: 'linode/debian8.1',
    vendor: 'Debian',
    label: 'Debian 8.1',
  },
  'linode/debian6': {
    ...testLinodeImage,
    id: 'linode/debian6',
    recommended: false,
    vendor: 'Debian',
    label: 'Debian 6',
  },
};

export const testPrivateImage = {
  ...testLinodeImage,
  id: 'private/38',
  label: 'testImage',
  description: 'image details',
  vendor: null,
  created: '2017-08-08T13:55:16',
  created_by: 'zues',
  size: 2048,
  type: 'manual',
  is_public: false,
};

export const privateImages = {
  'private/38': testPrivateImage,
  'private/39': {
    ...testPrivateImage,
    id: 'private/39',
    label: 'nextImage',
    description: 'image from deleted disk',
    created: '2017-08-08T13:55:16',
    created_by: 'zues',
    size: 2048,
    type: 'automatic',
    is_public: false,
  },
};

export const images = {
  ...privateImages,
  ...publicImages,
};
