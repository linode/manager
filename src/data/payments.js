export const apiTestPayment = {
  usd: 5.0,
  id: 1000,
  date: '2015-11-12T16:11:59',
};

export const payments = [0, 1, 2, 3, 4].map(i => ({
  ...apiTestPayment,
  id: apiTestPayment.id + i,
  usd: apiTestPayment.usd + i,
}));
