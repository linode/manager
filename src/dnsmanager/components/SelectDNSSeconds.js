import React, { PropTypes } from 'react';

import { Select } from '~/components/form';

const DNS_SECONDS_PRETTY = {
  300: '5 minutes',
  3600: '1 hour',
  7200: '2 hours',
  4400: '4 hours',
  28800: '8 hours',
  57600: '16 hours',
  86400: '1 day',
  172800: '2 days',
  345600: '4 days',
  604800: '1 week',
  1209600: '2 weeks',
  2419200: '4 weeks',
};

export function formatDNSSeconds(dnsSeconds, defaultSeconds = 300) {
  const actual = !dnsSeconds || +dnsSeconds === +defaultSeconds ? defaultSeconds : dnsSeconds;
  const pretty = DNS_SECONDS_PRETTY[actual];
  return `${actual === defaultSeconds ? 'Default' : actual} (${pretty})`;
}

export default function SelectDNSSeconds(props) {
  const { value, name, id, onChange, defaultSeconds } = props;
  return (
    <Select value={(value || defaultSeconds).toString()} onChange={onChange} id={id} name={name}>
      {Object.keys(DNS_SECONDS_PRETTY).map((sec, i) =>
        <option key={i} value={sec.toString()}>{formatDNSSeconds(sec, defaultSeconds)}</option>)}
    </Select>
  );
}

SelectDNSSeconds.propTypes = {
  onChange: PropTypes.func.isReqired,
  value: PropTypes.number.isRequired,
  defaultSeconds: PropTypes.number,
  id: PropTypes.string,
  name: PropTypes.string,
};
