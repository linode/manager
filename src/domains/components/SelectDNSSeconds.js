import React, { PropTypes } from 'react';

import { Select } from 'linode-components/forms';

export const FIVE_MINUTES = 300;
export const ONE_HOUR = FIVE_MINUTES * 12;
export const TWO_HOURS = ONE_HOUR * 2;
export const FOUR_HOURS = ONE_HOUR * 4;
export const EIGHT_HOURS = ONE_HOUR * 8;
export const SIXTEEN_HOURS = ONE_HOUR * 16;
export const ONE_DAY = ONE_HOUR * 24;
export const TWO_DAYS = ONE_DAY * 2;
export const FOUR_DAYS = ONE_DAY * 4;
export const ONE_WEEK = ONE_DAY * 7;
export const TWO_WEEKS = ONE_WEEK * 2;
export const FOUR_WEEKS = ONE_WEEK * 4;

const DNS_SECONDS_PRETTY = {
  [FIVE_MINUTES]: '5 minutes',
  [ONE_HOUR]: '1 hour',
  [TWO_HOURS]: '2 hours',
  [FOUR_HOURS]: '4 hours',
  [EIGHT_HOURS]: '8 hours',
  [SIXTEEN_HOURS]: '16 hours',
  [ONE_DAY]: '1 day',
  [TWO_DAYS]: '2 days',
  [FOUR_DAYS]: '4 days',
  [ONE_WEEK]: '1 week',
  [TWO_WEEKS]: '2 weeks',
  [FOUR_WEEKS]: '4 weeks',
};

export function formatDNSSeconds(dnsSeconds, defaultSeconds = ONE_DAY, withTitle = false) {
  const actualDefaultSeconds = defaultSeconds || ONE_DAY; // Takes care of if defaultSeconds is 0.
  const actual = !dnsSeconds || +dnsSeconds === +actualDefaultSeconds ?
                 actualDefaultSeconds : dnsSeconds;
  const pretty = DNS_SECONDS_PRETTY[actual];
  const actualOrDefault = actual === actualDefaultSeconds ? 'Default' : actual;

  if (withTitle) {
    return <span title={pretty}>{actualOrDefault}</span>;
  }

  return `${actualOrDefault}${pretty ? ` (${pretty})` : ''}`;
}

export default function SelectDNSSeconds(props) {
  const { value, name, id, onChange, defaultSeconds = ONE_DAY } = props;

  const actualValue = value || defaultSeconds || ONE_DAY; // Takes care of if defaultSeconds is 0.

  return (
    <Select value={actualValue.toString()} onChange={onChange} id={id} name={name}>
      {Object.keys(DNS_SECONDS_PRETTY).map((sec, i) =>
        <option key={i} value={sec.toString()}>{formatDNSSeconds(sec, defaultSeconds)}</option>)}
    </Select>
  );
}

SelectDNSSeconds.propTypes = {
  onChange: PropTypes.func.isRequired,
  value: PropTypes.any,
  defaultSeconds: PropTypes.number,
  id: PropTypes.string,
  name: PropTypes.string,
};
