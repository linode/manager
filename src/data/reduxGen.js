import { ONE, MANY, DELETE } from '~/api/internal.ts';

import { configsNodeBalancer, noGroupNodeBalancer } from '~/data/nodebalancers';

export const resource = { ...configsNodeBalancer };
export const page = {
  data: [configsNodeBalancer, noGroupNodeBalancer],
  pages: 1,
  results: 1,
  page: 1,
  nodebalancers: [configsNodeBalancer, noGroupNodeBalancer],
};

export const testConfigOne = {
  name: 'nodebalancers',
  primaryKey: 'id',
  endpoint: id => `/nodebalancers/${id}`,
  supports: [ONE],
};

export const testConfigMany = {
  name: 'nodebalancers',
  primaryKey: 'id',
  endpoint: id => `/nodebalancers/${id}`,
  supports: [MANY],
};

export const testConfigDelete = {
  name: 'nodebalancers',
  primaryKey: 'id',
  endpoint: id => `/nodebalancers/${id}`,
  supports: [DELETE],
};
