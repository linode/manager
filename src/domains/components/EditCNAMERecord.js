import React, { PropTypes, Component } from 'react';

import {
  Form,
  FormSummary,
  Input,
  SubmitButton,
  ModalFormGroup,
} from 'linode-components/forms';
import { CancelButton } from 'linode-components/buttons';

import { domains } from '~/api';
import { dispatchOrStoreErrors } from '~/api/util';

import SelectDNSSeconds from './SelectDNSSeconds';

export default class EditCNAMERecord extends Component {
  constructor(props) {
    super();

    const { id, zone: { ttl_sec: defaultTTL, domain: zone } } = props;
    const {
      target: alias,
      name: hostname,
      ttl_sec: ttl,
    } = props.zone._records.records[id] || {};

    this.state = {
      errors: {},
      loading: false,
      zone,
      defaultTTL,
      ttl,
      hostname,
      alias,
    };
  }

  onSubmit = () => {
    const { dispatch, id, close } = this.props;
    const { ttl, hostname, alias } = this.state;
    const ids = [this.props.zone.id, id].filter(Boolean);
    const data = {
      ttl_sec: +ttl,
      name: hostname,
      target: alias,
      type: 'CNAME',
    };

    return dispatch(dispatchOrStoreErrors.call(this, [
      () => domains.records[id ? 'put' : 'post'](data, ...ids),
      close,
    ]));
  }

  onChange = ({ target: { name, value } }) => this.setState({ [name]: value })

  render() {
    const { close } = this.props;
    const { errors, loading, defaultTTL, ttl, hostname, alias } = this.state;

    return (
      <Form onSubmit={this.onSubmit}>
        <ModalFormGroup errors={errors} id="hostname" label="Hostname" apiKey="name">
          <Input
            id="hostname"
            name="hostname"
            value={hostname}
            placeholder="www.thisdomain.com"
            onChange={this.onChange}
          />
        </ModalFormGroup>
        <ModalFormGroup label="Aliases to" id="alias" apiKey="target" errors={errors}>
          <Input
            id="alias"
            name="alias"
            value={alias}
            placeholder="www.otherdomain.com"
            onChange={this.onChange}
          />
        </ModalFormGroup>
        <ModalFormGroup label="TTL" id="ttl" apiKey="ttl_sec" errors={errors}>
          <SelectDNSSeconds
            id="ttl"
            name="ttl"
            value={ttl}
            defaultSeconds={defaultTTL}
            onChange={this.onChange}
          />
        </ModalFormGroup>
        <div className="Modal-footer">
          <CancelButton onClick={close} />
          <SubmitButton
            disabled={loading}
            disabledChildren={this.props.id ? undefined : 'Adding CNAME Record'}
          >
            {this.props.id ? undefined : 'Add CNAME Record'}
          </SubmitButton>
          <FormSummary errors={errors} />
        </div>
      </Form>
    );
  }
}

EditCNAMERecord.propTypes = {
  dispatch: PropTypes.func.isRequired,
  zone: PropTypes.object.isRequired,
  close: PropTypes.func.isRequired,
  id: PropTypes.number,
};
