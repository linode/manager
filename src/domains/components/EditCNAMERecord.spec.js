import { shallow } from 'enzyme';
import React from 'react';
import sinon from 'sinon';

import EditCNAMERecord from '~/domains/components/EditCNAMERecord';

import { createSimulatedEvent, expectDispatchOrStoreErrors, expectRequest } from '~/test.helpers';
import { api } from '~/data';
import { Input } from 'linode-components';
import SelectDNSSeconds from '~/domains/components/SelectDNSSeconds';

describe('domains/components/EditCNAMERecord', () => {
  const sandbox = sinon.sandbox.create();

  afterEach(() => {
    sandbox.restore();
  });

  it('renders fields correctly CNAME', () => {
    const currentZone = api.domains.domains['1'];
    const currentRecord = currentZone._records.records[2];
    const wrapper = shallow(
      <EditCNAMERecord
        dispatch={() => { }}
        zone={currentZone}
        id={currentRecord.id}
        close={() => { }}
      />
    );

    const hostname = wrapper.findWhere((a) => a.type() === Input
      && a.prop('id') === 'hostname');

    const alias = wrapper.findWhere((a) => a.type() === Input
      && a.prop('id') === 'alias');

    const ttl = wrapper.findWhere((a) => a.type() === SelectDNSSeconds
      && a.prop('id') === 'ttl');

    expect(hostname.props().value).toBe(currentRecord.name);
    expect(alias.props().value).toBe(currentRecord.target);
    expect(ttl.props().defaultSeconds).toBe(currentRecord.ttl_sec || currentZone.ttl_sec);
    expect(ttl.props().value).toBe(0);
  });

  it('saves an existing CNAME record', async () => {
    const dispatch = sandbox.stub();
    const currentZone = api.domains.domains['1'];
    const currentRecord = currentZone._records.records[4];
    const close = sandbox.spy();
    const wrapper = shallow(
      <EditCNAMERecord
        dispatch={dispatch}
        zone={currentZone}
        id={currentRecord.id}
        close={close}
      />
    );

    const hostname = wrapper.findWhere((a) => a.type() === Input
      && a.prop('id') === 'hostname');

    const alias = wrapper.findWhere((a) => a.type() === Input
      && a.prop('id') === 'alias');

    const ttl = wrapper.findWhere((a) => a.type() === SelectDNSSeconds
      && a.prop('id') === 'ttl');

    hostname.simulate('change', createSimulatedEvent('hostname', 'www.tester1234.com'));
    alias.simulate('change', createSimulatedEvent('alias', 'www.othertester1234.com'));
    ttl.simulate('change', createSimulatedEvent('ttl', 3600));

    await wrapper.props().onSubmit();

    expect(dispatch.callCount).toBe(1);
    await expectDispatchOrStoreErrors(dispatch.firstCall.args[0], [
      ([fn]) => expectRequest(fn, `/domains/${currentZone.id}/records/${currentRecord.id}`, {
        method: 'PUT',
        body: {
          name: 'www.tester1234.com',
          target: 'www.othertester1234.com',
          ttl_sec: 3600,
          type: 'CNAME',
        },
      }),
    ], 1);

    expect(close.callCount).toBe(1);
  });

  it('creates a new CNAME record', async () => {
    const dispatch = sandbox.stub();
    const currentZone = api.domains.domains['1'];
    const close = sandbox.spy();
    const wrapper = shallow(
      <EditCNAMERecord
        dispatch={dispatch}
        zone={currentZone}
        close={close}
      />
    );

    const hostname = wrapper.findWhere((a) => a.type() === Input
      && a.prop('id') === 'hostname');

    const alias = wrapper.findWhere((a) => a.type() === Input
      && a.prop('id') === 'alias');

    const ttl = wrapper.findWhere((a) => a.type() === SelectDNSSeconds
      && a.prop('id') === 'ttl');

    hostname.simulate('change', createSimulatedEvent('hostname', 'www.tester1234.com'));
    alias.simulate('change', createSimulatedEvent('alias', 'www.othertester1234.com'));
    ttl.simulate('change', createSimulatedEvent('ttl', 3600));


    await wrapper.props().onSubmit();

    expect(dispatch.callCount).toBe(1);
    await expectDispatchOrStoreErrors(dispatch.firstCall.args[0], [
      ([fn]) => expectRequest(fn, `/domains/${currentZone.id}/records/`, {
        method: 'POST',
        body: {
          name: 'www.tester1234.com',
          target: 'www.othertester1234.com',
          ttl_sec: 3600,
          type: 'CNAME',
        },
      }),
    ], 1);

    expect(close.callCount).toBe(1);
  });
});
