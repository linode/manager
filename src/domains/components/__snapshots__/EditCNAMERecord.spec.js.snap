// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`domains/components/EditCNAMERecord should render without error 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EditCNAMERecord
    close={[Function]}
    dispatch={[Function]}
    id={2}
    title="EditCNAMERecord"
    zone={
        Object {
            "_polling": false,
            "_records": Object {
              "records": Object {
                "1": Object {
                  "id": 1,
                  "name": "sub1",
                  "port": 80,
                  "priority": 10,
                  "protocol": null,
                  "service": null,
                  "target": "4.4.4.4",
                  "ttl_sec": 0,
                  "type": "A",
                  "weight": 5,
                },
                "2": Object {
                  "id": 2,
                  "name": "sub2",
                  "port": 80,
                  "priority": 10,
                  "protocol": null,
                  "service": null,
                  "target": "sub1.example.com",
                  "ttl_sec": 0,
                  "type": "CNAME",
                  "weight": 5,
                },
                "3": Object {
                  "id": 3,
                  "name": "",
                  "port": 0,
                  "priority": 10,
                  "protocol": null,
                  "service": null,
                  "target": "mail1.example.com",
                  "ttl_sec": 0,
                  "type": "MX",
                  "weight": 0,
                },
                "4": Object {
                  "id": 4,
                  "name": "",
                  "port": 0,
                  "priority": 0,
                  "protocol": null,
                  "service": null,
                  "target": "ns5.linode.com",
                  "ttl_sec": 0,
                  "type": "NS",
                  "weight": 0,
                },
                "5": Object {
                  "id": 5,
                  "name": "key",
                  "port": 80,
                  "priority": 10,
                  "protocol": null,
                  "service": null,
                  "target": "value",
                  "ttl_sec": 0,
                  "type": "TXT",
                  "weight": 5,
                },
                "6": Object {
                  "id": 6,
                  "name": "_sip",
                  "port": 80,
                  "priority": 10,
                  "protocol": "_tcp",
                  "service": "_sip",
                  "target": "ns1.service.com",
                  "ttl_sec": 0,
                  "type": "SRV",
                  "weight": 5,
                },
                "7": Object {
                  "id": 7,
                  "name": "www.mywebsite.com",
                  "tag": "issue",
                  "target": "certsign.ro",
                  "ttl_sec": 0,
                  "type": "CAA",
                },
              },
            },
            "axfr_ips": Array [
              "44.55.66.77",
            ],
            "description": "Example Description",
            "domain": "example.com",
            "expire_sec": 604800,
            "group": "Example Display Group",
            "id": 1,
            "master_ips": Array [
              "127.0.0.1",
              "255.255.255.1",
              "123.123.123.7",
            ],
            "refresh_sec": 14400,
            "retry_sec": 3600,
            "soa_email": "admin@example.com",
            "status": "active",
            "ttl_sec": 3600,
            "type": "master",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "analytics": Object {
        "action": "edit",
        "title": "EditCNAMERecord",
      },
      "buttonDisabledText": "Saving",
      "buttonText": "Save",
      "children": <div>
        <ModalFormGroup
                apiKey="name"
                errors={Object {}}
                id="hostname"
                label="Hostname"
        >
                <Input
                        className=""
                        id="hostname"
                        name="hostname"
                        onChange={[Function]}
                        placeholder="www"
                        value="sub2"
                />
        </ModalFormGroup>
        <ModalFormGroup
                apiKey="target"
                errors={Object {}}
                id="alias"
                label="Aliases to"
        >
                <Input
                        className=""
                        id="alias"
                        name="alias"
                        onChange={[Function]}
                        placeholder="www.otherdomain.com"
                        value="sub1.example.com"
                />
        </ModalFormGroup>
        <ModalFormGroup
                apiKey="ttl_sec"
                errors={Object {}}
                id="ttl"
                label="TTL"
        >
                <SelectDNSSeconds
                        defaultSeconds={3600}
                        id="ttl"
                        name="ttl"
                        onChange={[Function]}
                        value={0}
                />
        </ModalFormGroup>
</div>,
      "errors": Object {},
      "noCancel": false,
      "noCancelEvent": false,
      "noSubmit": false,
      "onCancel": [Function],
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <ModalFormGroup
            apiKey="name"
            errors={Object {}}
            id="hostname"
            label="Hostname"
>
            <Input
                        className=""
                        id="hostname"
                        name="hostname"
                        onChange={[Function]}
                        placeholder="www"
                        value="sub2"
            />
</ModalFormGroup>,
          <ModalFormGroup
            apiKey="target"
            errors={Object {}}
            id="alias"
            label="Aliases to"
>
            <Input
                        className=""
                        id="alias"
                        name="alias"
                        onChange={[Function]}
                        placeholder="www.otherdomain.com"
                        value="sub1.example.com"
            />
</ModalFormGroup>,
          <ModalFormGroup
            apiKey="ttl_sec"
            errors={Object {}}
            id="ttl"
            label="TTL"
>
            <SelectDNSSeconds
                        defaultSeconds={3600}
                        id="ttl"
                        name="ttl"
                        onChange={[Function]}
                        value={0}
            />
</ModalFormGroup>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "apiKey": "name",
            "children": <Input
              className=""
              id="hostname"
              name="hostname"
              onChange={[Function]}
              placeholder="www"
              value="sub2"
/>,
            "errors": Object {},
            "id": "hostname",
            "label": "Hostname",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "className": "",
              "id": "hostname",
              "name": "hostname",
              "onChange": [Function],
              "placeholder": "www",
              "value": "sub2",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "apiKey": "target",
            "children": <Input
              className=""
              id="alias"
              name="alias"
              onChange={[Function]}
              placeholder="www.otherdomain.com"
              value="sub1.example.com"
/>,
            "errors": Object {},
            "id": "alias",
            "label": "Aliases to",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "className": "",
              "id": "alias",
              "name": "alias",
              "onChange": [Function],
              "placeholder": "www.otherdomain.com",
              "value": "sub1.example.com",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "apiKey": "ttl_sec",
            "children": <SelectDNSSeconds
              defaultSeconds={3600}
              id="ttl"
              name="ttl"
              onChange={[Function]}
              value={0}
/>,
            "errors": Object {},
            "id": "ttl",
            "label": "TTL",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "defaultSeconds": 3600,
              "id": "ttl",
              "name": "ttl",
              "onChange": [Function],
              "value": 0,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": "div",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "analytics": Object {
          "action": "edit",
          "title": "EditCNAMERecord",
        },
        "buttonDisabledText": "Saving",
        "buttonText": "Save",
        "children": <div>
          <ModalFormGroup
                    apiKey="name"
                    errors={Object {}}
                    id="hostname"
                    label="Hostname"
          >
                    <Input
                              className=""
                              id="hostname"
                              name="hostname"
                              onChange={[Function]}
                              placeholder="www"
                              value="sub2"
                    />
          </ModalFormGroup>
          <ModalFormGroup
                    apiKey="target"
                    errors={Object {}}
                    id="alias"
                    label="Aliases to"
          >
                    <Input
                              className=""
                              id="alias"
                              name="alias"
                              onChange={[Function]}
                              placeholder="www.otherdomain.com"
                              value="sub1.example.com"
                    />
          </ModalFormGroup>
          <ModalFormGroup
                    apiKey="ttl_sec"
                    errors={Object {}}
                    id="ttl"
                    label="TTL"
          >
                    <SelectDNSSeconds
                              defaultSeconds={3600}
                              id="ttl"
                              name="ttl"
                              onChange={[Function]}
                              value={0}
                    />
          </ModalFormGroup>
</div>,
        "errors": Object {},
        "noCancel": false,
        "noCancelEvent": false,
        "noSubmit": false,
        "onCancel": [Function],
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <ModalFormGroup
              apiKey="name"
              errors={Object {}}
              id="hostname"
              label="Hostname"
>
              <Input
                            className=""
                            id="hostname"
                            name="hostname"
                            onChange={[Function]}
                            placeholder="www"
                            value="sub2"
              />
</ModalFormGroup>,
            <ModalFormGroup
              apiKey="target"
              errors={Object {}}
              id="alias"
              label="Aliases to"
>
              <Input
                            className=""
                            id="alias"
                            name="alias"
                            onChange={[Function]}
                            placeholder="www.otherdomain.com"
                            value="sub1.example.com"
              />
</ModalFormGroup>,
            <ModalFormGroup
              apiKey="ttl_sec"
              errors={Object {}}
              id="ttl"
              label="TTL"
>
              <SelectDNSSeconds
                            defaultSeconds={3600}
                            id="ttl"
                            name="ttl"
                            onChange={[Function]}
                            value={0}
              />
</ModalFormGroup>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "apiKey": "name",
              "children": <Input
                className=""
                id="hostname"
                name="hostname"
                onChange={[Function]}
                placeholder="www"
                value="sub2"
/>,
              "errors": Object {},
              "id": "hostname",
              "label": "Hostname",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "className": "",
                "id": "hostname",
                "name": "hostname",
                "onChange": [Function],
                "placeholder": "www",
                "value": "sub2",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "apiKey": "target",
              "children": <Input
                className=""
                id="alias"
                name="alias"
                onChange={[Function]}
                placeholder="www.otherdomain.com"
                value="sub1.example.com"
/>,
              "errors": Object {},
              "id": "alias",
              "label": "Aliases to",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "className": "",
                "id": "alias",
                "name": "alias",
                "onChange": [Function],
                "placeholder": "www.otherdomain.com",
                "value": "sub1.example.com",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "apiKey": "ttl_sec",
              "children": <SelectDNSSeconds
                defaultSeconds={3600}
                id="ttl"
                name="ttl"
                onChange={[Function]}
                value={0}
/>,
              "errors": Object {},
              "id": "ttl",
              "label": "TTL",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "defaultSeconds": 3600,
                "id": "ttl",
                "name": "ttl",
                "onChange": [Function],
                "value": 0,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": "div",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
